video/out/vo_drm.c

```diff
@@ -0,0 +1,523 @@
+            }
+            MP_ERR(vo, "Cannot create framebuffer for connector %d\n",
+                   conn_id);
+            return ret;
+        }
+    }
+
+end:
+    if (conn) {
+        drmModeFreeConnector(conn);
+        conn = NULL;
+    }
+    if (res) {
+        drmModeFreeResources(res);
+        res = NULL;
+    }
+    if (ret == 0) {
+        *out = dev;
+    } else {
+        talloc_free(dev);
+    }
+    return ret;
+}
+
+
+
+static int reconfig(struct vo *vo, struct mp_image_params *params, int flags)
+{
+    struct priv *p = vo->priv;
+    float device_ar = p->device_w / (float)p->device_h;
+
+    vo_get_src_dst_rects(vo, &p->src, &p->dst, &p->osd);
+    int32_t video_w = (p->dst.x1 - p->dst.x0);
+    int32_t video_h = (p->dst.y1 - p->dst.y0);
+    float video_ar = video_w / (float)video_h;
+    int32_t w, h;
+    if (device_ar > video_ar) {
+        w = p->device_h * video_ar;
+        h = p->device_h;
+    } else {
+        w = p->device_w;
+        h = p->device_w * video_ar;
+    }
+    if (w > p->device_w) w = p->device_w;
+    if (h > p->device_h) h = p->device_h;
+
+    // p->osd contains the parameters assuming OSD rendering in window
+    // coordinates, but OSD can only be rendered in the intersection
+    // between window and video rectangle (i.e. not into panscan borders).
+    p->osd.w = w;
+    p->osd.h = h;
+    p->osd.mt = MPMIN(0, p->osd.mt);
+    p->osd.mb = MPMIN(0, p->osd.mb);
+    p->osd.mr = MPMIN(0, p->osd.mr);
+    p->osd.ml = MPMIN(0, p->osd.ml);
+
+    p->x = (p->device_w - w) >> 1;
+    p->y = (p->device_h - h) >> 1;
+    p->buf = (char*)talloc_size(vo, p->device_w * p->device_h * 4);
+    memset(p->buf, 0, p->device_w * p->device_h * 4);
+
+    mp_sws_set_from_cmdline(p->sws, vo->opts->sws_opts);
+    p->sws->src = *params;
+    p->sws->dst = (struct mp_image_params) {
+        .imgfmt = IMGFMT_BGR0,
+        .w = w,
+        .h = h,
+        .d_w = w,
+        .d_h = h,
+    };
+
+    MP_INFO(vo, "resizing %dx%d -> %dx%d\n",
+            p->sws->src.w,
+            p->sws->src.h,
+            p->sws->dst.w,
+            p->sws->dst.h);
+
+    mp_image_params_guess_csp(&p->sws->dst);
+
+    if (mp_sws_reinit(p->sws) < 0)
+        return 1;
+
+    vo->want_redraw = true;
+    return 0;
+}
+
+static void draw_image(struct vo *vo, mp_image_t *mpi)
+{
+    struct priv *p = vo->priv;
+
+    struct modeset_buf *front_buf = &p->dev->bufs[p->dev->front_buf];
+    struct mp_image img = {0};
+    mp_image_set_params(&img, &p->sws->dst);
+    img.planes[0] = p->buf + p->x * 4;
+    img.stride[0] = front_buf->stride;
+
+    struct mp_rect src_rc = p->src;
+    src_rc.x0 = MP_ALIGN_DOWN(src_rc.x0, mpi->fmt.align_x);
+    src_rc.y0 = MP_ALIGN_DOWN(src_rc.y0, mpi->fmt.align_y);
+    mp_image_crop_rc(mpi, src_rc);
+
+    mp_sws_scale(p->sws, &img, mpi);
+
+    osd_draw_on_image(vo->osd, p->osd, mpi ? mpi->pts : 0, 0, &img);
+
+    memmove(front_buf->map, p->buf, p->device_w * p->device_h * 4);
```

Comment:
I've read somewhere that `memmove()` operates faster but I don't remember why. If I had to guess, `memmove` can be translated directly to `rep movsb`, while `memcpy` doesn't allow that because of its well documented side effects (i.e. the overlapping bytes), so it gets translated to a simple loop. But after testing it thoroughly I see `gcc` translates both `memcpy` and `memmove` into `rep movsl`, so this micro-optimization makes no sense.